### ISSUE TITLE
Support to convert CudnnLSTM params to normal Weights and Bias

### ISSUE THREAD

**Issue by robotnc** [tensorflow/tensorflow#6072]
My environment is Tensorflow r0.11, Linux 16.04, CUDA8.0, cuDNN5.1

I am using the CudnnLSTM class  from tensorflow.contrib.cudnn_rnn , the training speed is really fast with about 5~6x speedup. However after training I need to move the model to an embeded system which is not CUDA based. So how can I convert the CudnnLSTM params to normal weights and bias. 

Is this feature still in plan or already have other solutions to this issue ?

**Comment by zheng-xq** [comment]
Yes, it is being worked on. Adding zhangyaobit@.

**Comment by zhangyaobit** [comment]
@robotnc Yes, we have checked in https://github.com/tensorflow/tensorflow/commit/ebcf0ee4e7e5541ffc418ca832223c409b9a2e57

You can use RNNParamsSaveable and your customized function canonical_to_params to restore the previously saved cuDNN parameters the way you need.

Please let us know how it goes. This will be very cool!

**Comment by zhangyaobit** [comment]
Adding ebrevdo@, who is interested in this.

**Comment by robotnc** [comment]
@zhangyaobit Yeah, it looks cool. I think I need some time to recompile and have a try, because I am not so familiar with how to patch,compile and etc.

And one more question, I see the function below:

+  def params_to_canonical(self, params):
 +    """Converts params from a specific format of cuDNN to the canonical format.
 +
 +    Args:
 +      params: a Variable for weight and bias parameters.
 +
 +    Returns:
 +      A function for the specific-to-canonical conversion.
 +    """
 +    weights, biases = gen_cudnn_rnn_ops.cudnn_rnn_params_to_canonical(
 +        num_layers=self._num_layers,
 +        num_units=self._num_units,
 +        input_size=self._input_size,
 +        params=params,
 +        num_params=self._num_layers * self._NUM_PARAMS_PER_LAYER,
 +        rnn_mode=self._rnn_mode,
 +        input_mode=self._input_mode,
 +        direction=self._direction)
 +    return weights, biases

If the RNN is multi-layer, how can I get the weights and bias for each layer? Because I need to convert the model to an embedded system where a new model will be  re-contructed with each layer's weights and bias.

**Comment by zhangyaobit** [comment]
@robotnc, as you see, the return values of params_to_canonical are weights and biases; they are stored in a layer-by-layer order.

For example, for LSTM, there are 8 matrices in each layer and the first 8 weight matrices stored in weights belong to the first layer; out of these 8 matrices, the first 4 matrices are applied to the input from the previous layer, and the second 4 matrices are applied to the recurrent input. Biases are stored in the same order.

You can write your canonical_to_params to take in weights and biases and output whatever data format you want for your model.

Please let me if anything is not clear.

**Comment by robotnc** [comment]
@zhangyaobit I success to run the code, and hope to know more detail about the weights structure.
My model is two layer LSTM with 128cells each, now I use params_to_cannoical(params) get the weights. The weights include 8 tensors inside, which I can see it like this:

  weights = [W0,W1,W2,W3, ... W15], total 16 tensors in list

I assume that the first 8 tensors are for layer-0, and the last 8 tensors are for layer-1.
However in each layer, we have 8 matrixes which are 

  W_f,W_o,W_c,W_i,R_f,R_o,R_c,R_i (f: forget gate, o: output gate, c: saved status, i: input gate)

So the question is what is the correct order(f, o, c, i) of these matrixes?
Finally I need to convert the weights to be compatible with tf.nn.rnn_cell.BasicLSTMCell.

![cudnn_lstm](https://cloud.githubusercontent.com/assets/18507467/21179737/5ff2233c-c230-11e6-8b57-6a73e92dfab0.png)

**Comment by zhangyaobit** [comment]
I believe this is the answer you are looking for. 

Excerpt from "4.105. cudnnGetRNNLinLayerMatrixParams" in cuDNN guide:

If mode in rnnDesc was set to CUDNN_LSTM values of 0-3
reference matrix multiplications applied to the input from the
previous layer, value of 4-7 reference matrix multiplications
applied to the recurrent input.
‣ Values 0 and 4 reference the input gate.
‣ Values 1 and 5 reference the forget gate.
‣ Values 2 and 6 reference the new memory gate.
‣ Values 3 and 7 reference the output gate.

**Comment by robotnc** [comment]
That’s great! Thanks a lot!

**Comment by robotnc** [comment]
@zhangyaobit I met another issue in Cudnn_LSTM about dropout, it seems the dropout didn't take effect. I have raised a new issue here, can you check it for me ?
https://github.com/tensorflow/tensorflow/issues/6466

**Comment by boche** [comment]
If a layer is bidirectional, how are the 16 (8 * 2) matrices ordered ?

**Comment by zhangyaobit** [comment]
I don't think being bidirectional affect the order; the same order is used. 

Could you clarify what 16, 8, and 2 respectively mean here?

**Comment by boche** [comment]
@zhangyaobit 
If unidirectional, one layer has 8 matrices ordered like W_i,W_f,W_c,W_o,R_i,R_f,R_c,R_o.
If bidirectional, one layer will then have 16 matrices right, since we have a forward matrix and a backward matrix for each W_x or R_x, so my question is how are these 16 matrices ordered?

It can be appended like W^f_i, ...,R^f_o, W^b_i, ..., R^b_o
or interleaved as W^f_i, W^b_i, ..., R^f_o, R^b_o
^f: stand for forward matrix, ^b stand for backward matrix

I am not sure which way is it organized (appending form, interleaving form, or some other form), and which direction is put first (forward first or backward first). Hope this clarifies my question.

**Comment by boche** [comment]
Also I found a bug @zhangyaobit. I am using params_to_canonical to export parameters.

When my model is bidirectional, I always have this mistake

> F tensorflow/contrib/cudnn_rnn/kernels/cudnn_rnn_ops.cc:550] Check failed: size == width * height Params size mismatch. Expected 100, got 1290

Here my input size is 129, num_units is 10. But when I switch my model to unidirectional, there is no such mistake.

I think the bug is in line 548.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/kernels/cudnn_rnn_ops.cc#L548-L550

The logic of this line is to use input_size as width for the first half parameters (W matrix), and num_units as width for the second half parameters (R matrix). But in bidirectional case, the param is probably not organized in the order of first W and then R, for example it can be W^f_x, R^f_x, W^b_x, R^b_x (f: forward, b: backward, x: stand for different gates).

**Comment by zhangyaobit** [comment]
@boche Thanks for reporting the issue. Could you file a separate github issue on this? As the original issue of this thread has been resolved, I'm closing it now. Thanks.

I will double check if bidirectional RNN is currently supported. Looks like not yet. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cudnn_rnn/python/kernel_tests/cudnn_rnn_ops_test.py#L148

**Comment by hyichao** [comment]
Hi @robotnc @zhangyaobit    

It's still confusing for me that how cudnn lstm put the weights.   
My case is a layer of cudnn lstm with input 1024-dim feauture and hidden state size is 512.
The shape of cudnn lstm param matrix is  (512, 6152)

According to the previous comments, I notice that the order of gates are [I,F,C,O], but how is the order of weights and bias? [W W W W W W W W B B B B B B B B] or [W B W B W B W B W B W B W B W B ] ?

If it is like [Wi, Wf, Wc, Wo, Ri, Rf, Rc, Ro, Bwi, Bwf, Bwc, Bwo, Bri, Brf, Brc, Bro],
then the codes should be
 
```
        input_size = 1024
        hidden_size = 512     
        input_w_ifco = trained_weights[:,  0 :4*input_size]
        hidden_w_ifco = trained_weights[:, 4*input_size: 4*input_size + 4*hidden_size]
        input_b_ifco = trained_weights[:, 4*input_size + 4*hidden_size: 4*input_size + 4*hidden_size + 4]
        hidden_b_ifco = trained_weights[:, 4*input_size + 4*hidden_size + 4 : 4*input_size + 4*hidden_size + 8]
```

Or if it is like [Wi, Wf, Wc, Wo, Bwi, Bwf, Bwc, Bwo, Ri, Rf, Rc, Ro, Bri, Brf, Brc, Bro]

```
        input_size = 1024
        hidden_size = 512
        input_w_ifco = trained_weights[:,  0 :4*input_size]
        input_b_ifco = trained_weights[:, 4*input_size: 4*input_size+ 4]
        hidden_w_ifco = trained_weights[:, 4*input_size + 4: 4*input_size + 4 + 4*hidden_size]
        hidden_b_ifco = trained_weights[:, 4*input_size + +4 +4*hidden_size : 4*input_size + + 4 + 4*hidden_size + 4]
```

Many thanks !
